compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -2405,6 +2438,34 @@ fn add_upstream_rust_crates<'a>(
                 // external build system already has the native dependencies defined, and it
                 // will provide them to the linker itself.
                 if sess.opts.unstable_opts.link_native_libraries {
+                    if sess.opts.unstable_opts.separate_native_rlib_dependencies {
+                        // if rlib contains native libs as archieves, unpack them to tmpdir
+                        let archive_map = unsafe {
+                            Mmap::map(File::open(&src.rlib.as_ref().unwrap().0).unwrap()).unwrap()
+                        };
+                        let archive =
+                            object::read::archive::ArchiveFile::parse(&*archive_map).unwrap();
+                        for entry in archive.members() {
+                            if let Ok(entry) = entry {
+                                let begin: usize = entry.file_range().0.try_into().unwrap();
+                                let end: usize = entry.file_range().1.try_into().unwrap();
+                                let end = end + begin;
+                                let data = &archive_map[begin..end];
+
+                                let mut name_buf = Vec::from(entry.name().clone());
+                                name_buf.push(0);
+                                let name: &[u8] = &name_buf;
+                                let name = core::ffi::CStr::from_bytes_with_nul(name).unwrap();
+                                let name = name.to_str().unwrap();
+                                if !name.to_string().ends_with(".a") {
+                                    continue;
+                                }
+                                let name = PathBuf::from(name.clone());
+                                fs::write(&tmpdir.join(name), data).unwrap();
```

Comment:
See `fn get_rlib_metadata` and `fn add_archive` for an examples of doing all this stuff above in a nicer way.